// @tenna-llc/hapi-plugin-utilities/test/plugin/mocks/options.js
'use strict';

module.exports = {
    serverStopTimeoutMilliseconds: 5000,
    paths: {
        health: '/health',
        topLevelVersion: '/version'
    },
    clients: {
        bunnyBus: {
            'us-east-1': {
                hostname: '127.0.0.1',
                port: '15672',
                username: 'guest',
                password: 'guest'
            }
        },
        awsDynamoDB: {
            'us-east-1': {
                region: 'us-east-1',
                endpoint: 'http://localhost:8000',
                credentials: {
                    accessKeyId: 'abc',
                    secretAccessKey: 'def'
                }
            }
        },
        awsS3: {
            'us-east-1': {
                region: 'us-east-1',
                endpoint: 'http://localhost:4566',
                credentials: {
                    accessKeyId: 'abc',
                    secretAccessKey: 'def'
                },
                sslEnabled: true,
                s3ForcePathStyle: true
            }
        },
        redis: {
            'us-east-1': {
                host: 'localhost',
                port: '6379'
            }
        },
        redisCluster: {
            'us-east-1': {
                nodes: [],
                options: { maxRedirections: 1 }
            }
        },
        sequelize: {
            someDB: {
                dialect: 'postgres',
                username: 'postgres',
                password: 'postgres',
                database: 'postgres',
                host: 'localhost',
                port: '5432',
                pool: {
                    min: 0,
                    max: 5,
                    idle: 10000
                }
            }
        },
        superAgent: {
            crud: {
                host: 'dv3.nonprod.tenna.com',
                prefix: '/api/v4'
            }
        }
    }
};
